### ISSUE TITLE
GPU not detected inside Pod

### ISSUE THREAD

**Issue by ZafeerCodes** [NexusGPU/tensor-fusion#313]
**Issue:** I deployed a sample pod, it went  into running state, but unable to detect GPU 

- Baremetal kubernetes setup
- Ubuntu 22.04
- NVIDIA  RTX 5090

` k get tensorfusioncluster
NAME            PHASE     TOTAL POOLS   TOTAL NODES   TOTAL GPU   TOTAL TFLOPS   TOTAL VRAM   AVAILABLE TFLOPS   AVAILABLE VRAM
tensor-fusion   Running   1             1             1           165160m        24564Mi      165160m            24564Mi
`

`kubectl get pods -n tensor-fusion-sys
NAME                                            READY   STATUS      RESTARTS   AGE
hypervisor-baseer                               2/2     Running     0          13h
node-discovery-baseer-w7cgw                     0/1     Completed   0          3h40m
tensor-fusion-sys-controller-55fd5f5b97-8n47w   2/2     Running     0          13h
`

**Pod Descripition**

` k describe po pytorch-example-6c58f8c77c-tzltm
Name:             pytorch-example-6c58f8c77c-tzltm
Namespace:        default
Priority:         0
Service Account:  default
Node:             baseer/192.168.0.46
Start Time:       Thu, 07 Aug 2025 00:56:50 +0530
Labels:           app=pytorch-example
                  pod-template-hash=6c58f8c77c
                  tensor-fusion.ai/component=client
                  tensor-fusion.ai/enabled=true
                  tensor-fusion.ai/gpupool=tensor-fusion-shared
                  tensor-fusion.ai/pod-template-hash=65dd2f2bf920a0c2
Annotations:      cni.projectcalico.org/containerID: 8fa86adbb39554581b73369f061c8ed04bbd48cd1b7c657e14e0957190bb1441
                  cni.projectcalico.org/podIP: 192.168.12.143/32
                  cni.projectcalico.org/podIPs: 192.168.12.143/32
                  tensor-fusion.ai/generate-workload: true
                  tensor-fusion.ai/gpu-count: 1
                  tensor-fusion.ai/gpupool: tensor-fusion-shared
                  tensor-fusion.ai/inject-container: python
                  tensor-fusion.ai/is-local-gpu: false
                  tensor-fusion.ai/qos: medium
                  tensor-fusion.ai/replicas: 1
                  tensor-fusion.ai/selected-workload: pytorch-example
                  tensor-fusion.ai/tflops-limit: 10
                  tensor-fusion.ai/tflops-request: 20
                  tensor-fusion.ai/vram-limit: 4Gi
                  tensor-fusion.ai/vram-request: 4Gi
                  tensor-fusion.ai/workload: pytorch-example
Status:           Running
IP:               192.168.12.143
IPs:
  IP:           192.168.12.143
Controlled By:  ReplicaSet/pytorch-example-6c58f8c77c
Init Containers:
  inject-lib:
    Container ID:   cri-o://c57d6b03f8658b8db6f4d52c35dae391624056fbc9f41178c545af338a00d20f
    Image:          tensorfusion/tensor-fusion-client:1.47.0
    Image ID:       docker.io/tensorfusion/tensor-fusion-client@sha256:2d8c4fe04b7c9996b85ec60bc8aee10434d73b5ae7f69c3c01fa40d7257fd93e
    Port:           <none>
    Host Port:      <none>
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 07 Aug 2025 00:56:50 +0530
      Finished:     Thu, 07 Aug 2025 00:56:50 +0530
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /tensor-fusion from tf-libs (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-lsdsq (ro)
Containers:
  python:
    Container ID:  cri-o://15ec03becc9ab28153956e5220f461d22de2a3c9a97bd31cbd57f1c92eaa9650
    Image:         pytorch/pytorch:2.6.0-cuda12.4-cudnn9-runtime
    Image ID:      docker.io/pytorch/pytorch@sha256:77f17f843507062875ce8be2a6f76aa6aa3df7f9ef1e31d9d7432f4b0f563dee
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      sleep 1d
    State:          Running
      Started:      Thu, 07 Aug 2025 01:09:11 +0530
    Ready:          True
    Restart Count:  0
    Environment:
      TF_ENABLE_LOG:                              1
      TF_LOG_PATH:                                /tmp/tf.log
      TF_PREPEND_PATH:                            /tensor-fusion
      TF_PREPEND_LD_LIBRARY_PATH:                 /tensor-fusion
      TENSOR_FUSION_CONNECTION_NAME:              pytorch-example-6c58f8c77c-1sdbh1hvrr
      TENSOR_FUSION_CONNECTION_NAMESPACE:         default
      TENSOR_FUSION_OPERATOR_GET_CONNECTION_URL:  http://tensor-fusion-sys.tensor-fusion-sys:8080/api/connection?name=pytorch-example-6c58f8c77c-1sdbh1hvrr&namespace=default
    Mounts:
      /etc/ld.so.preload from tf-libs (ro,path="ld.so.preload")
      /tensor-fusion from tf-libs (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-lsdsq (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  kube-api-access-lsdsq:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    Optional:                false
    DownwardAPI:             true
  tf-libs:
    Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:      
    SizeLimit:   <unset>
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:          <none>
`

**Stuck here:**
`
kubectl exec -it pytorch-example-6c58f8c77c-tzltm -- /bin/sh
Defaulted container "python" out of: python, inject-lib (init)
# python3
Python 3.11.11 | packaged by conda-forge | (main, Dec  5 2024, 14:17:24) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> print("CUDA available:", torch.cuda.is_available())
`

`nvidia-device-plugin-daemonset-h27tm       1/1     Running   0                46h
`

`baseer   Ready    control-plane   47h   v1.31.4   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=baseer,kubernetes.io/os=linux,node-role.kubernetes.io/control-plane=,nvidia.com/gpu.present=true
`

#/etc/crio/crio.conf

`which nvidia-container-runtime
/usr/bin/nvidia-container-runtime
(base) baseer@baseer:~$ cat /etc/crio/crio.conf

[crio]

  [crio.image]
    insecure_registries = ["192.168.0.46:6000", "14.194.41.182:6060"]

  [crio.runtime]

    [crio.runtime.runtimes]

      [crio.runtime.runtimes.nvidia]
        runtime_path = "/usr/bin/nvidia-container-runtime"
        runtime_type = "oci"

`

Please let me know if you need some other details.

**Comment by Code2Life** [comment]
Hi @ZafeerCodes  could you paste the `torch.cuda.device_count()` result ?  
there was another issue but not related to CUDA missing, pytorch 2.6 can not be run on 50 series GPU, need to install pytorch with '--pre' for pytorch pre release.

**Comment by Code2Life** [comment]
Resolved in discord, @ZafeerCodes thanks for trying and using TensorFusion
